### PR TITLE
Os Upgrade Feature

### PR DESCRIPTION
**Pull request by shivamverma-ms**
## Description

- Added -OsUpgradeVersion Parameter to the Start-Migration and Start-TestMigration
- Added the respective field in Input.csv

Az version: `10.0.1`

### CODE REVIEW

File: azure-migrate/migrate-at-scale-vmware-agentles/Agentless VMware automation/AzMigrate_StartMigration.ps1 [hunk 1]
```diff
@@ -64,13 +64,19 @@ Function ProcessItemImpl($processor, $csvItem, $reportItem) {
...
+    if([string]::IsNullOrEmpty($osUpgradeVersion)){
+        $processor.Logger.LogTrace("OS_UPGRADE_VERSION is not mentioned for: '$($sourceMachineName)'")
+        $reportItem.AdditionalInformation = "OS_VERSION_UPGRADE is not mentioned for: '$($sourceMachineName)'"
+    }
+
     #start the migration
     if ([string]::IsNullOrEmpty($TurnOff_SrcServer) -or ($TurnOff_SrcServer -eq 'N') -or ($TurnOff_SrcServer -eq 'No')){
         #we are defaulting to this if Turn off Source Server is not mentioned
-        $MigrateJob = Start-AzMigrateServerMigration -InputObject $ReplicatingServermachine
+        $MigrateJob = Start-AzMigrateServerMigration -InputObject $ReplicatingServermachine -OsUpgradeVersion $osUpgradeVersion
```
Comment (singhabh27) [inline azure-migrate/migrate-at-scale-vmware-agentles/Agentless VMware automation/AzMigrate_StartMigration.ps1:76]: We should pass $osUpgradeVersion only if it is not null or empty.

File: azure-migrate/migrate-at-scale-vmware-agentles/Agentless VMware automation/AzMigrate_StartTestMigration.ps1 [hunk 2]
```diff
@@ -105,7 +105,13 @@ Function ProcessItemImpl($processor, $csvItem, $reportItem) {
...
     #End Code for Target Subscription
 
-    $TestMigrationJob = Start-AzMigrateTestMigration -InputObject $ReplicatingServermachine -TestNetworkID $Target_VNet.Id
+    $osUpgradeVersion = $csvItem.OS_UPGRADE_VERSION
+    if([string]::IsNullOrEmpty($osUpgradeVersion)){
+        $processor.Logger.LogTrace("OS_UPGRADE_VERSION is not mentioned for: '$($sourceMachineName)'")
+        $reportItem.AdditionalInformation = "OS_VERSION_UPGRADE is not mentioned for: '$($sourceMachineName)'"
+    } 
+
+    $TestMigrationJob = Start-AzMigrateTestMigration -InputObject $ReplicatingServermachine -TestNetworkID $Target_VNet.Id -OsUpgradeVersion $osUpgradeVersion
```
Comment (singhabh27) [inline azure-migrate/migrate-at-scale-vmware-agentles/Agentless VMware automation/AzMigrate_StartTestMigration.ps1:114]: $osUpgradeVersion should not be passed always.

File: azure-migrate/migrate-at-scale-vmware-agentles/Agentless VMware automation/AzMigrate_UpdateMachineProperties.ps1 [hunk 3]
```diff
@@ -235,6 +235,76 @@ Function ProcessItemImpl($processor, $csvItem, $reportItem) {
     } 
     #End Code for Target Subscription
 
+    $targetDiskName = $csvItem.UPDATED_TARGET_DISK_NAME
```
Comment (singhabh27) [inline azure-migrate/migrate-at-scale-vmware-agentles/Agentless VMware automation/AzMigrate_UpdateMachineProperties.ps1:238]: Remove custom resource naming changes from this PR.

File: azure-migrate/migrate-at-scale-vmware-agentles/Agentless VMware automation/AzMigrate_UpdateMachineProperties.ps1 [hunk 4]
```diff
@@ -303,6 +303,7 @@ Function ProcessItemImpl($processor, $csvItem, $reportItem) {
     else {
         $paramsNIC1.Add("NicId", $UpdatedNIC1ID)
     }
+
```
Comment (singhabh27) [inline azure-migrate/migrate-at-scale-vmware-agentles/Agentless VMware automation/AzMigrate_UpdateMachineProperties.ps1:306]: nit: Remove extra lines.

File: azure-migrate/migrate-at-scale-vmware-agentles/Agentless VMware automation/AzMigrate_UpdateMachineProperties.ps1 [hunk 5]
```diff
@@ -347,6 +348,7 @@ Function ProcessItemImpl($processor, $csvItem, $reportItem) {
     else {
         $paramsNIC2.Add("NicId", $UpdatedNIC2ID)
     }
+
     $NIC2_SelectionType = $csvItem.UPDATED_TARGET_NIC2_SELECTIONTYPE
```
Comment (singhabh27) [inline azure-migrate/migrate-at-scale-vmware-agentles/Agentless VMware automation/AzMigrate_UpdateMachineProperties.ps1:352]: nit: Remove extra lines

File: azure-migrate/migrate-at-scale-vmware-agentles/Agentless VMware automation/AzMigrate_StartMigration.ps1 [hunk 6]
```diff
@@ -64,13 +64,27 @@ Function ProcessItemImpl($processor, $csvItem, $reportItem) {
...
+            $MigrateJob = Start-AzMigrateServerMigration -InputObject $ReplicatingServermachine -OsUpgradeVersion $osUpgradeVersion
+        }
     }
     else {
-        $MigrateJob = Start-AzMigrateServerMigration -InputObject $ReplicatingServermachine -TurnOffSourceServer
+        if([string]::IsNullOrEmpty($osUpgradeVersion)){
+            $processor.Logger.LogTrace("OS_UPGRADE_VERSION is not mentioned for: '$($sourceMachineName)'")
+            $reportItem.AdditionalInformation = "OS_VERSION_UPGRADE is not mentioned for: '$($sourceMachineName)'"
+            $MigrateJob = Start-AzMigrateServerMigration -InputObject $ReplicatingServermachine -TurnOffSourceServer
+        } else {
```
Comment (singhabh27) [inline azure-migrate/migrate-at-scale-vmware-agentles/Agentless VMware automation/AzMigrate_StartMigration.ps1:85]: nit: else in next line.

**Comment by shivamverma-ms** [comment]
#sign-off

**Review by mikefrobbins — APPROVED** [review]
LGTM

**Pull request merged**